File: mcp-code-intelligence-nodejs/src/memory/session-repo.ts
/**
 * SessionRepository — tracks MCP sessions (one per connection).
 */

import Database from 'better-sqlite3';
import { randomUUID } from 'crypto';
import { MemorySession } from './models.js';

export class SessionRepository {
  private readonly db: Database.Database;

  constructor(db: Database.Database) {
    this.db = db;
  }

  /** Start a new session, returns session ID (8-char UUID prefix). */
  startSession(agentName?: string): string {
    const sessionId = randomUUID().slice(0, 8);
    this.db.prepare(
      'INSERT INTO memory_sessions (session_id, agent_name) VALUES (?, ?)'
    ).run(sessionId, agentName ?? null);
    return sessionId;
  }

  /** End a session. */
  endSession(sessionId: string): void {
    this.db.prepare(
      "UPDATE memory_sessions SET status = 'ended', ended_at = datetime('now') WHERE session_id = ?"
    ).run(sessionId);
  }

  /** Increment observation count. */
  incrementObservations(sessionId: string): void {
    this.db.prepare(
      'UPDATE memory_sessions SET observation_count = observation_count + 1 WHERE session_id = ?'
    ).run(sessionId);
  }

  /** List recent sessions. */
  listRecent(limit = 20): MemorySession[] {
    return this.db.prepare(
      'SELECT * FROM memory_sessions ORDER BY started_at DESC LIMIT ?'
    ).all(limit) as MemorySession[];
  }

  /** List sessions with optional agent and status filters. */
  listFiltered(agent: string, status: string, limit: number): MemorySession[] {
    const clauses: string[] = ['1=1'];
    const params: unknown[] = [];
    if (agent) { clauses.push('agent_name = ?'); params.push(agent); }
    if (status) { clauses.push('status = ?'); params.push(status); }
    params.push(limit);
    const where = clauses.join(' AND ');
    return this.db.prepare(
      `SELECT * FROM memory_sessions WHERE ${where} ORDER BY started_at DESC LIMIT ?`
    ).all(...params) as MemorySession[];
  }

  /** Get active session count. */
  activeCount(): number {
    const row = this.db.prepare(
      "SELECT COUNT(*) as cnt FROM memory_sessions WHERE status = 'active'"
    ).get() as { cnt: number };
    return row.cnt;
  }
}

File: mcp-code-intelligence-python/src/mcp_code_intel/memory/session_repo.py
"""SessionRepository — tracks MCP sessions (one per connection)."""

import sqlite3
import uuid
from typing import Any


class SessionRepository:
    """Session tracking for memory engine."""

    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def start_session(self, agent_name: str | None = None) -> str:
        """Start a new session, returns 8-char session ID."""
        session_id = uuid.uuid4().hex[:8]
        self._conn.execute(
            "INSERT INTO memory_sessions (session_id, agent_name) VALUES (?, ?)",
            (session_id, agent_name),
        )
        self._conn.commit()
        return session_id

    def end_session(self, session_id: str) -> None:
        """End a session."""
        self._conn.execute(
            "UPDATE memory_sessions SET status = 'ended', ended_at = datetime('now') WHERE session_id = ?",
            (session_id,),
        )
        self._conn.commit()

    def increment_observations(self, session_id: str) -> None:
        """Increment observation count."""
        self._conn.execute(
            "UPDATE memory_sessions SET observation_count = observation_count + 1 WHERE session_id = ?",
            (session_id,),
        )
        self._conn.commit()

    def list_recent(self, limit: int = 20) -> list[dict[str, Any]]:
        """List recent sessions."""
        cur = self._conn.execute(
            "SELECT * FROM memory_sessions ORDER BY started_at DESC LIMIT ?",
            (limit,),
        )
        return [dict(r) for r in cur.fetchall()]

    def active_count(self) -> int:
        """Get active session count."""
        cur = self._conn.execute(
            "SELECT COUNT(*) FROM memory_sessions WHERE status = 'active'"
        )
        return cur.fetchone()[0]
